src/daemon/emitter.cc

```diff
@@ -527,5 +582,63 @@ void Emitter::DoRemoteExecute(const base::WorkerPool& pool,
   }
 }
 
+void Emitter::DoCoordinator(const base::WorkerPool& pool,
+                            std::vector<ResolveFn> resolvers) {
+  net::EndPointPtr end_point;
+  net::ConnectionPtr connection;
+  ui32 sleep_period = 1;
+  auto Sleep = [&sleep_period]() mutable {
+    LOG(INFO) << "Sleeping for " << sleep_period
+              << " seconds before next attempt";
+    std::this_thread::sleep_for(std::chrono::seconds(sleep_period));
+    if (sleep_period < static_cast<ui32>(-1) / 2) {
+      sleep_period <<= 1;
+    }
+  };
+
+  compilation_workers_.reset(new base::WorkerPool(true));
+  while (!pool.IsShuttingDown()) {
+    if (!connection) {
+      for (const ResolveFn& resolver : resolvers) {
+        end_point = resolver();
+        if (!end_point)
+          continue;
+
+        String error;
+        connection = Connect(end_point, &error);
+        if (!connection) {
+          LOG(WARNING) << "Failed to connect to " << end_point->Print() << ": "
+                       << error;
+          continue;
+        }
+        break;
+      }
+    }
+    if (!connection) {
+      Sleep();
+      continue;
+    }
+
+    sleep_period = 1;
+    UniquePtr<proto::Configuration> configuration(new proto::Configuration);
+    if (!connection->SendSync(std::move(configuration)))
+      continue;
+
+    Universal reply(new net::proto::Universal);
+    if (!connection->ReadSync(reply.get()))
+      continue;
+
+    if (reply->HasExtension(proto::Configuration::extension)) {
+      auto* config = reply->MutableExtension(proto::Configuration::extension);
+      UpdateCompilationPool(*config);
```

Comment:
I tried to use that method, but it resulted as overcomplicated code. It's called from `Initialize` - what for? So it could be called on different threads. As a result more difficult pool terminating mechanics to make sure only thread safe methods are used.